### PR TITLE
Introduce state management package to keep state of messages feature and global state

### PR DESCRIPTION
**Pull request by gilbertsoft**
## The Issue

State is currently mixed into configuration but should be separated into a dedicated state file instead. E.g. there is some state in the globalconfig `~/.ddev/global_config.yml` namely the project infos or last started version.

## How This PR Solves The Issue

This PR introduces a small state management package which makes state handling easy possible by defining e.g. a state struct. The package supports the management of multiple states completey independant, means no cross dependencies are added from a state to another, a single state works completely independant. This ensures a clean per package state management but in a central place for the whole application within one state file. A state is accessed by his unique key.

* this package will be used by #4885 which e.g. needs to know about the last update or the last message which was shown.
* also a PR will follow which decouples the state from the globalconfig e.g. project infos or last started version which should not be stored in the config but a separate state file.

PR should be merged as soon as the tests are completed.

## Manual Testing Instructions

No manual testing is possible at the moment because the code is not used so far.

## Automated Testing Overview

The packages are fully automated tested with multiple unit tests and for the YAML state storage also benchmark tests are included.

## Release/Deployment Notes

No side effects are possible at the moment because the code is not used so far.

<a href="https://gitpod.io/#https://github.com/ddev/ddev/pull/4948"></a>

### CODE REVIEW

**Comment by rfay** [comment]
Please add to the OP a practical statement about what this will be used for and what it means. You'll need to explain "state management" and its impact on existing code.

Also, please say whether this could be introduced in the near term or whether it needs to wait until after our major release.

**Comment by gilbertsoft** [comment]
> Please add to the OP a practical statement about what this will be used for and what it means. You'll need to explain "state management" and its impact on existing code.
> 
> Also, please say whether this could be introduced in the near term or whether it needs to wait until after our major release.

Done. As written above code is currently not in use but a preparation for #4885.

**Comment by rfay** [comment]
What I'm asking for is a practical description of what this does and why, and how it will change other code. Please choose an example command/state/config and explain how this will change that.

**Comment by gilbertsoft** [comment]
> What I'm asking for is a practical description of what this does and why, and how it will change other code. Please choose an example command/state/config and explain how this will change that.

sth like this? https://github.com/ddev/ddev/pull/4948/files#diff-102188c33aa91417b0ad17f55fa004a97b79d743024b7be5c1db269da3380076

**Comment by rfay** [comment]
I'm asking you to improve the PR OP above with a practical example, taken from DDEV, of how this changes things. Explain what this means to DDEV, what it changes, what motivates it.

**Comment by rfay** [comment]
https://github.com/mitchellh/mapstructure looks like a wonderful thing. Could be used several other places.

File: pkg/config/state/state_manager_test.go
Comment (rfay) [inline pkg/config/state/state_manager_test.go:145]: Do these test suite items need to have global scope or should this be testSetPreservesExistingStates()?

File: pkg/config/state/storage/yaml/yaml_storage.go
Comment (rfay) [inline pkg/config/state/storage/yaml/yaml_storage.go:15]: Wow could this be nice. So many places.

**Review by rfay — APPROVED** [review]
Thanks, I'll go with it! You can merge when you think it's ready. I usually try to make sure that the tests on master have succeeded before merging something new; the last two should be getting there pretty soon, but that latest commit there isn't very important (tests only) and the last two tests don't even use mutagen.

File: pkg/config/state/storage/yaml/yaml_storage.go
Comment (gilbertsoft) [reply pkg/config/state/storage/yaml/yaml_storage.go:15]: > https://github.com/mitchellh/mapstructure looks like a wonderful thing. Could be used several other places.

Yes, makes some things easier and more flexible compared to the json and yaml implementation.

**Comment by gilbertsoft** [comment]
> https://github.com/mitchellh/mapstructure looks like a wonderful thing. Could be used several other places.

Yes, makes some things easier and more flexible compared to the json and yaml implementation.

File: pkg/config/state/storage/yaml/yaml_storage.go
Comment (gilbertsoft) [reply pkg/config/state/storage/yaml/yaml_storage.go:15]: He he, factory functions :) Easier to use with strict typed parameters compared to a manual setup of the struct. And you also can call methods e.g. or do other important init stuff.

**Pull request merged**
